From 0841bbf69ed2b7d623c09a8e35ddf563a2acfb18 Mon Sep 17 00:00:00 2001
From: Matteo Croce <teknoraver@meta.com>
Date: Fri, 18 Oct 2024 18:44:13 +0200
Subject: [PATCH] keep on using DBus as fallback if varlink is not available

---
 src/network/networkctl.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/network/networkctl.c b/src/network/networkctl.c
index a447c39..f3d7f24 100644
--- a/src/network/networkctl.c
+++ b/src/network/networkctl.c
@@ -173,7 +173,7 @@ int acquire_bus(sd_bus **ret) {
         if (networkd_is_running()) {
                 r = varlink_connect_networkd(/* ret_varlink = */ NULL);
                 if (r < 0)
-                        return r;
+                        log_warning("Varlink connection failed, fallback to D-Bus.");
         } else
                 log_warning("systemd-networkd is not running, output might be incomplete.");
 
-- 
2.47.0

