diff -Naur a/src/test/meson.build b/src/test/meson.build
--- a/src/test/meson.build	2021-02-02 07:29:47.000000000 -0800
+++ b/src/test/meson.build	2021-02-19 10:58:19.188785027 -0800
@@ -61,11 +61,6 @@
           libmount,
           libblkid]],
 
-        [['src/test/test-emergency-action.c'],
-         [libcore,
-          libshared],
-         []],
-
         [['src/test/test-chown-rec.c'],
          [libcore,
           libshared],
@@ -522,24 +517,6 @@
          [],
          []],
 
-        [['src/test/test-tables.c',
-          'src/shared/test-tables.h',
-          'src/journal/journald-server.c',
-          'src/journal/journald-server.h'],
-         [libcore,
-          libjournal_core,
-          libudev_core,
-          libudev_static,
-          libsystemd_network,
-          libshared],
-         [threads,
-          libseccomp,
-          libmount,
-          libxz,
-          liblz4,
-          libblkid],
-         '', '', [], libudev_core_includes],
-
         [['src/test/test-prioq.c'],
          [],
          []],
@@ -621,11 +598,6 @@
          [],
          []],
 
-        [['src/test/test-cgroup-cpu.c'],
-         [libcore,
-          libshared],
-         []],
-
         [['src/test/test-cgroup-unit-default.c'],
          [libcore,
           libshared],
@@ -1168,10 +1140,6 @@
          [],
          []],
 
-        [['src/analyze/test-verify.c', 'src/analyze/analyze-verify.c', 'src/analyze/analyze-verify.h'],
-         [libcore, libshared],
-         []],
-
         [['src/login/test-inhibit.c'],
          [],
          [],
