From 69443affcaa711a9a3355994458542d3732631a5 Mon Sep 17 00:00:00 2001
From: Anita Zhang <the.anitazha@gmail.com>
Date: Tue, 30 Nov 2021 12:08:50 -0800
Subject: [PATCH] Re-enable repart if libfdisk ver. has range fix

---
 meson.build | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/meson.build b/meson.build
index 6263e7c0fc..35e6c4f376 100644
--- a/meson.build
+++ b/meson.build
@@ -1011,7 +1011,7 @@ libmount = dependency('mount',
 want_libfdisk = get_option('fdisk')
 if want_libfdisk != 'false' and not skip_deps
         libfdisk = dependency('fdisk',
-                              version : '>= 2.33',
+                              version : '>= 2.32.1',
                               required : want_libfdisk == 'true')
         have = libfdisk.found()
 else
--
2.31.1

